File: contiguoussum.py
# Time Complexity : O(N^2)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Ran for 69/80 test cases. Time limit exceeded for the rest.
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach
# Used brute force to iterate over the array and find if each subarray sum is equal to k

class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        n = len(nums)
        count = 0
        if n == 1:
            if k == nums[0]:
                return 1
            else:
                return 0
        for i in range(0,n):
            sum = 0
            for j in range(i,n):
                if i == j:
                    sum += nums[i]
                    if nums[i] == k:
                        count +=1
                        
                else:
                    sum += nums[j]
                    if sum == k:
                        count +=1
        return count
                    
        

File: ContiguousSum.java
// Time Complexity : O(N), where N = length of array
// Space Complexity : O(N).
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No.

class Solution {
    public int findMaxLength(int[] nums) {
        // Map to store first index this sum has occured
        Map<Integer, Integer> sumToIndexMap = new HashMap<>();
        // '0' can be obtained from empty array as well, so 0's default index is -1
        sumToIndexMap.put(0, -1);
        int rSum = 0, len = 0, maxLen = 0;
        for(int i = 0; i < nums.length; i++)
        {
            // if it is 1 add 1; or else if it is 0 add -1.
            rSum += nums[i] == 1 ? 1 : -1;

            // check if Running sum was already in map
            // if it was that means this is balanced Subarray from prevIndex till this
            if(sumToIndexMap.containsKey(rSum))
            {
                len = i - sumToIndexMap.get(rSum);
                maxLen = Math.max(len, maxLen);
            }
            else
            {
                // update map
                sumToIndexMap.put(rSum, i);
            }
        }
        
        return maxLen;
    }
}